### ISSUE TITLE
? in assignment trigger set_next_input 

### ISSUE THREAD

**Issue by Carreau** [ipython/ipython#10226]
See https://github.com/jupyter/notebook/issues/2107

**Comment by takluyver** [comment]
It's a feature, not a bug!

Seriously, I added this deliberately years ago, so that when I'm typing `a = foo.bar.some_long_name(baz`, I can add a `?` to see info on `baz` and carry on right where I was. It makes sense in the terminal, but not so much in the notebook, where we tend to use shift-tab in that situation. It seemingly doesn't get used very often, which is why people don't notice it.

We could call `set_next_input(..., replace=True)`, so that in the notebook it replaces the cell rather than  adding a new one. Or we could only do it in terminal IPython, not in the kernel.

**Comment by Nagasaki45** [comment]
but why

```python
a = foo.bar.some_long_name(baz?
```

should behave differently from

```python
foo.bar.some_long_name(baz?
```
?

In case it's an intended behaviour, I was expecting it to be the same in these two examples.

**Comment by takluyver** [comment]
Did you try it? At least for me, it is the same in those two cases.

This behaviour is triggered when there's something other than names and attribute access (`foo.bar`) in the input. This may seem somewhat arbitrary, but it works out quite nicely in the terminal; if I've just done `foo.bar?`, I've typically written the whole line to inspect that thing, but if I do `foo.bar.some_long_name(baz?`, I'm probably in the middle of doing something when I decided to inspect something.

**Comment by Nagasaki45** [comment]
> Did you try it? At least for me, it is the same in those two cases.

You are right. What I actually tried is the example presented in the original ticket (jupyter/notebook#2107), but in IPython.

> This behaviour is triggered when there's something other than names and attribute access (foo.bar) in the input.

Thanks for the clarification! I didn't understand these differences before your last comment. IMHO it still feels a bit arbitrary, even inside the terminal...

**Comment by smartsammler** [comment]
I'm not sure if this is also the cause for the following strange behaviour I stumbled upon when debugging a similar problem like the one of the issue opener:
``` Python
In [1]: "".replace?
Object `replace` not found.
In [2]: "".format?
Signature: format(value, format_spec='', /)
Docstring:
Return value.__format__(format_spec)

format_spec defaults to the empty string
Type:      builtin_function_or_method
```

and on the other side:
``` Python
In [1]: a = ""
In [2]: a.replace?
Docstring:
S.replace(old, new[, count]) -> str

Return a copy of S with all occurrences of substring
old replaced by new.  If the optional argument count is
given, only the first count occurrences are replaced.
Type:      builtin_function_or_method
In [3]: a.format?
Docstring:
S.format(*args, **kwargs) -> str

Return a formatted version of S, using substitutions from args and kwargs.
The substitutions are identified by braces ('{' and '}').
Type:      builtin_function_or_method
```
So the second ones behave like using `help(str.` and show the docstrings of the methods of the `str` class. Whereas the first behaviour seems to work on objects and have got problems with objects that are not in saved in variables (here `""`)

In my opinion the behaviour is strange (a) because having a variable and not having it yet show different results and (b) because class methods, like `replace`, behave different from class methods with the same name as global functions, like `format`. I would prefer it if both usages would behave like in the second case.

**Comment by takluyver** [comment]
That's a totally separate matter. It would be nice if `"".replace?` worked, but I think it's not easy with our current introspection system.

**Comment by smartsammler** [comment]
Ok, thanks for the reply.
